services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,276 @@
+  if (client_state_ == kStopped) {
+    return;
+  }
+
+  client_state_ = kStopped;
+
+  // Invalidate weak pointers, so we don't continue any async operations.
+  weak_ptr_factory_.InvalidateWeakPtrs();
+
+  // Stop can be called before the client_ has resolved.
+  if (!client_) {
+    return;
+  }
+
+  client_->Stop();
+
+  // Reset the client.
+  client_.reset();
+}
+
+const mojom::GeopositionResult* GeoClueLocationProvider::GetPosition() {
+  return last_position_.get();
+}
+
+void GeoClueLocationProvider::OnPermissionGranted() {
+  permission_granted_ = true;
+  MaybeStartClient();
+}
+
+void GeoClueLocationProvider::SetPosition(
+    mojom::GeopositionResultPtr position) {
+  last_position_ = std::move(position);
+
+  if (client_state_ != kStarted) {
+    return;
+  }
+
+  if (last_position_->is_position() &&
+      device::ValidateGeoposition(*last_position_->get_position())) {
+    position_update_callback_.Run(this, last_position_.Clone());
+    return;
+  }
+}
+
+void GeoClueLocationProvider::OnGetClientCompleted(
+    std::unique_ptr<GeoClueClientObject> client) {
+  if (!client) {
+    SetPosition(GetError());
+    return;
+  }
+
+  client_ = std::move(client);
+  SetClientProperties();
+}
+
+void GeoClueLocationProvider::SetClientProperties() {
+  auto* properties = client_->properties();
+
+  // Wait for all properties to be set before starting the client.
+  const auto callback = base::BarrierCallback<bool>(
+      2, base::BindOnce(&GeoClueLocationProvider::OnSetClientProperties,
+                        weak_ptr_factory_.GetWeakPtr()));
+
+  properties->requested_accuracy_level.Set(
+      high_accuracy_requested_ ? GeoClueClientObject::AccuracyLevel::kExact
+                               : GeoClueClientObject::AccuracyLevel::kCity,
+      callback);
+
+  properties->desktop_id.Set(kBraveDesktopId, callback);
+}
+
+void GeoClueLocationProvider::OnSetClientProperties(std::vector<bool> success) {
+  if (base::ranges::any_of(success, [](const auto& value) { return !value; })) {
+    VLOG(1) << "Failed to set properties. GeoClue2 location provider will "
+               "not work properly";
+    SetPosition(GetError());
+    return;
+  }
+
+  ConnectSignal();
+}
+
+void GeoClueLocationProvider::ConnectSignal() {
+  CHECK(client_);
+
+  client_->ConnectToLocationUpdatedSignal(
+      base::BindRepeating(&GeoClueLocationProvider::OnReadGeoClueLocation,
+                          weak_ptr_factory_.GetWeakPtr()),
+      base::BindOnce(&GeoClueLocationProvider::OnSignalConnected,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void GeoClueLocationProvider::OnSignalConnected(
+    const std::string& interface_name,
+    const std::string& signal_name,
+    bool success) {
+  if (!success) {
+    VLOG(1) << "Failed to connect to LocationUpdated Signal. GeoClue2 "
+               "location provider will "
+               "not work properly";
+    SetPosition(GetError());
+    return;
```

Comment:
Previously, it would freeze in whichever state it would fail in, which was fine for our purposes. In the refactor of GeoClueClientObject, I've added an error state to make things more obvious.